File: Python/6py 2/sq_in_rect.py
"""
|--------------------------------------------------------------------------
| Rectangle into Squares
|--------------------------------------------------------------------------
|
"""

def sq_in_rect(lng, wdth):
   if lng == wdth:
      return None
   squares = []
   while lng != 0 and wdth != 0:
      if lng > wdth:
         squares.append(wdth)
         lng -= wdth
      else:
         squares.append(lng)
         wdth -= lng
   return squares


def sqInRect(lng, wdth, ret = []):
   return None if lng==wdth else ret if not wdth else ([wdth]*(lng//wdth)) +  sqInRect(wdth, lng%wdth, ret) 


def sqInRect(a, b):
   if a == b: return None
   if not b: return []
   return [b] * (a // b) + sqInRect(b, a % b)


print(sq_in_rect(5, 5), None)
print(sq_in_rect(5, 3), [3, 2, 1, 1])
print(sq_in_rect(3, 5), [3, 2, 1, 1])
print(sq_in_rect(20, 14), [14, 6, 6, 2, 2, 2])

File: JavaScript/6js 2/sqInRect.js
/*
|--------------------------------------------------------------------------
| Rectangle into Squares
|--------------------------------------------------------------------------
|
*/

function sqInRect(lng, wdth) {
   if (lng === wdth) {
      return null;
   }
   let squares = [];
   while (lng !== 0 && wdth !== 0) {
      if (lng > wdth) {
         squares.push(wdth);
         lng -= wdth;
      } else {
         squares.push(lng);
         wdth -= lng;
      }
   }
   return squares;
}


function sqInRect(a, b, initial = true) {
   if (a === b) {
      return initial ? null : [a];
   }
   const min = Math.min(a, b);
   const max = Math.max(a, b);
   return [min, ...sqInRect(max - min, min, false)];
}


function sqInRect(width, height) {
   if (width === height) return null;
   if (width < height) [width, height] = [height, width];
   if (width === 2 * height) return [height, height];
   return [height, ...sqInRect(height, width - height)];
}


console.log(sqInRect(5, 5), null);
console.log(sqInRect(5, 3), [3, 2, 1, 1]);
console.log(sqInRect(3, 5), [3, 2, 1, 1]);
console.log(sqInRect(20, 14), [14, 6, 6, 2, 2, 2]);